### PR TITLE
NextupN function

### PR DESCRIPTION
**Pull request by pdrocaldeira**
nextupN implementation of TS 18661-2.

It's a nextafter but always towards positive infinity

Related to #94

### CODE REVIEW

**Comment by pmur** [comment]
I recommend squashing these into a single patch.

File: dfp/math.h [hunk 1]
```diff
@@ -265,6 +265,15 @@ extern _Decimal32 __rintd32 (_Decimal32 __x) __THROW;
...
 extern _Decimal32 __nextafterd32 (_Decimal32 __x, _Decimal32 __y) __THROW __attribute__ ((__const__));
 
+extern _Decimal32 nextupd32 (_Decimal32 __x) __THROW __attribute__ ((__const__));
+extern _Decimal32 __nextupd32 (_Decimal32 __x) __THROW __attribute__ ((__const__));
+
+extern _Decimal64 nextupd64 (_Decimal64 __x) __THROW __attribute__ ((__const__));
+extern _Decimal64 __nextupd64 (_Decimal64 __x) __THROW __attribute__ ((__const__));
+
+extern _Decimal128 nextupd128 (_Decimal128 __x) __THROW __attribute__ ((__const__));
+extern _Decimal128 __nextupd128 (_Decimal128 __x) __THROW __attribute__ ((__const__));
```
Comment (pmur) [inline dfp/math.h]: The TS 18661-2 prototypes should be guarded by the definition of `__STDC_WANT_IEC_60559_DFP_EXT__`.

File: ieee754r/nextupd64.c [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/* Returns the next representable neighbor of x in the direction of y
+
+   Copyright (C) 2006 IBM Corporation.
+   Copyright (C) 2007-2015 Free Software Foundation, Inc.
```
Comment (pmur) [inline ieee754r/nextupd64.c]: This should be updated to 2019. I don't think the IBM copyright is relevant to newly added files.

File: ieee754r/nextupd64.c [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+   This file is part of the Decimal Floating Point C Library.
+
+   Author(s): Joseph Kerian <jkerian@us.ibm.com>
```
Comment (pmur) [inline ieee754r/nextupd64.c]: Likewise.

File: tests/nextup.input [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# name nextup
+# arg1 decimal
+# ret  decimal
+
+0                                        0.000001 decimal32
+-0                                       0.000001 decimal32
+1                                        1.000001 decimal32
+2                                        2.000001 decimal32
+3                                        3.000001 decimal32
```
Comment (pmur) [inline tests/nextup.input]: I would also add a test for the smallest possible negative value, and -0.

File: ieee754r/nextupd32.c [hunk 5]
```diff
@@ -0,0 +1,58 @@
+/* Returns the next representable neighbor of x in the direction of plus infinity
+
+   Copyright (C) 2006 IBM Corporation.
+   Copyright (C) 2007-2015 Free Software Foundation, Inc.
```
Comment (pmur) [inline ieee754r/nextupd32.c]: The copyright and authors' should be updated.

File: ieee754r/nextupd128.c [hunk 6]
```diff
@@ -0,0 +1,29 @@
+/* Returns the next representable neighbor of x in the direction of y
+
+   Copyright (C) 2006 IBM Corporation.
+   Copyright (C) 2007-2015 Free Software Foundation, Inc.
```
Comment (pmur) [inline ieee754r/nextupd128.c]: Author and copyright should also be fixed up here too.

File: tests/nextup.input (hunk 4 above)
Comment (pmur) [reply tests/nextup.input]: The existing test suite can also be used to substitute interesting values for each type (e.g infinite, min/max values).

File: ieee754r/nextupd128.c (hunk 6 above)
Comment (pdrocaldeira) [reply ieee754r/nextupd128.c]: https://github.com/libdfp/libdfp/blob/21b9650b3c8015da74ef360cb86111864c08dc91/ieee754r/nextupd128.c#L7

Comment (pdrocaldeira) [reply ieee754r/nextupd128.c]: https://github.com/libdfp/libdfp/blob/21b9650b3c8015da74ef360cb86111864c08dc91/ieee754r/nextupd128.c#L3-L7

File: tests/nextup.input (hunk 4 above)
Comment (pdrocaldeira) [reply tests/nextup.input]: Now I'm using the test suite constants but I think some of them are wrong. For example, `DEC_MAX_EXP` is replaced by 90/369/6111 for 32/64/128 bits respectively and I think it should be 97/385/6145 based on ISO TR-24732.

There's negative 0 and smallest negative test for every type.

File: ieee754r/nextupd64.c (hunk 3 above)
Comment (pdrocaldeira) [reply ieee754r/nextupd64.c]: https://github.com/libdfp/libdfp/blob/21b9650b3c8015da74ef360cb86111864c08dc91/ieee754r/nextupd64.c#L3-L7

File: ieee754r/nextupd64.c (hunk 2 above)
Comment (pdrocaldeira) [reply ieee754r/nextupd64.c]: https://github.com/libdfp/libdfp/blob/21b9650b3c8015da74ef360cb86111864c08dc91/ieee754r/nextupd64.c#L3-L7

File: dfp/math.h (hunk 1 above)
Comment (pdrocaldeira) [reply dfp/math.h]: Please give a look on this commit: https://github.com/libdfp/libdfp/pull/95/commits/88f0f10cf9f73578b901cdf52ce0f03a5066bb84

I've changed the Makefile to define this constant at compilation time. I think I'm cheating but I don't know how to do it in another way.

File: tests/nextup.input (hunk 4 above)
Comment (pmur) [reply tests/nextup.input]: The maximum exponent is dissimilar to BFP. The mantissa is always an integral value, thus the maximum exponent of the format is reduced depending on the precision of the format. The maximum representable exponent is a value with one significand leading the radix point e.g 1.000eX.

Comment (pdrocaldeira) [reply tests/nextup.input]: Oh, now it makes sense. Thanks!

File: ieee754r/nextupd32.c [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+DEC_TYPE
+INTERNAL_FUNCTION_NAME (DEC_TYPE x)
+{
+	  DEC_TYPE z = IEEE_FUNCTION_NAME (x);
+	    if (!FUNC_D(__isfinite) (z) && FUNC_D(__isfinite) (x))
+		        DFP_ERRNO (ERANGE);
+	      return z;
+}
```
Comment (pmur) [inline ieee754r/nextupd32.c]: Errno should not be modified by nextup or nextdown.

Comment (pdrocaldeira) [reply ieee754r/nextupd32.c]: You mean, like that?
https://github.com/libdfp/libdfp/pull/95/commits/9b3d5c9d6b804cb033008ec107613f737345afb9

**Comment by pmur** [comment]
Would it be possible to merge the nextup and nextdown into a single PR with 2 patches?

**Comment by pdrocaldeira** [comment]
This code is now on #112
